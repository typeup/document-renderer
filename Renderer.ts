import * as dom from "@typeup/dom"
export type renderFunction = (renderer: Renderer, node: dom.Node) => Promise<string>

export class Renderer {
	private constructor(private variables: { [name: string]: string } = {}) {
	}
	getVariable(name: string): string {
		return this.variables[name]
	}
	setVariable(name: string, value: string) {
		this.variables[name] = value
	}
	async render(node: dom.Node | dom.Node[]): Promise<string> {
		let result: string
		if (Array.isArray(node))
			result = (await Promise.all(node.map(n => this.render(n)))).reduce<string[]>((r, e) => Array.isArray(e) ? [...r, ...e] : [...r, e], []).join("")
		else {
			const render = renderers[node.class]
			result = render ? await render(this, node) : ""
		}
		return result
	}
	scope(): Renderer {
		return new Renderer({ ...this.variables })
	}
	static create(): Renderer {
		return new Renderer({})
	}
}
const renderers: { [className: string]: renderFunction } = {}
export function addRenderer(className: string, render: renderFunction) {
	renderers[className] = render
}
