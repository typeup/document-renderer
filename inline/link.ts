import * as dom from "@typeup/dom"
import { Renderer, addRenderer } from "../Renderer"

async function render(renderer: Renderer, me: dom.inline.Link): Promise<string> {
	return `<a href='${me.target}'>${ await renderer.render(me.content) }</a>`
}
addRenderer("Inline.Link", render)
