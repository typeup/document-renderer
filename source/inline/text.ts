import * as dom from "@typeup/dom"
import { Renderer, addRenderer } from "../Renderer"

async function render(_: Renderer, me: dom.inline.Text): Promise<string> {
	return me.value
}
addRenderer("Inline.Link", render)
