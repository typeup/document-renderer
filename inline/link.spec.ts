import * as dom from "@typeup/dom"
import { Renderer } from "../Renderer"
import "./link"
import "./text"

const renderer = Renderer.create()
describe("Inline.Link", () => {
	it("render", async () => {
		expect(await renderer.render(new dom.inline.Link("http://example.com", [new dom.inline.Text("link")]))).toMatchSnapshot()
	})
})
