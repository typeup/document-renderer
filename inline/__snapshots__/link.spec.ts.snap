// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inline.Link render 1`] = `"<a href='http://example.com'>link</a>"`;
